pandas/core/generic.py

```diff
@@ -8180,6 +8180,90 @@ def ranker(data):
 
         return ranker(data)
 
+    _shared_docs[
+        "differences"
+    ] = """
+        Compare to another %(klass)s and show the differences.
+
+        The axis on which to stack results and how much information to
+        preserve can be customized.
+
+        Note that NaNs are considered not different from other NaNs.
+
+        Parameters
+        ----------
+        other : %(klass)s
+            Object to compare with.
+
+        axis : {0 or 'index', 1 or 'columns'}, default 1
+            Determine how the differences are stacked.
+            * 0, or 'index' : Stack differences on neighbouring rows.
+            * 1, or 'columns' : Stack differences on neighbouring columns.
+
+        keep_indices : bool, default False
+            Whether to keep the rows and columns that are equal, or drop them.
+
+        keep_values : bool, default False
+            Whether to keep the values that are equal, or show as NaNs.
+        """
+
+    @Appender(_shared_docs["differences"] % _shared_doc_kwargs)
+    def differences(self, other, axis=1, keep_indices=False, keep_values=False):
+        from pandas.core.reshape.concat import concat
+
+        mask = ~((self == other) | (self.isna() & other.isna()))
+        keys = ["self", "other"]
+
+        if not keep_values:
+            self = self.where(mask)
+            other = other.where(mask)
+
+        if not keep_indices:
+            if isinstance(self, ABCDataFrame):
+                cmask = mask.any()
+                rmask = mask.any(axis=1)
+                self = self.loc[rmask, cmask]
+                other = other.loc[rmask, cmask]
+            else:
+                self = self[mask]
+                other = other[mask]
+
+        if axis in (1, "columns"):  # This is needed for Series
+            axis = 1
+        else:
+            axis = self._get_axis_number(axis)
+
+        diff = concat([self, other], axis=axis, keys=keys)
+
+        if axis >= self.ndim:
+            # No need to reorganize data if stacking on new axis
+            # This currently applies for stacking two Series on columns
+            return diff
+
+        ax = diff._get_axis(axis)
+        ax_names = np.array(ax.names)
```

Comment:
why are you doing this?

Reply:
This is to deal with the same issue as #30485. I want to reorder the index levels according to their position, and not the index names. For example, if we stack the results on `axis='columns'`, we want to reorder the levels from `[0, 1]` to `[1, 0]` after we do concatenation with `keys=["self", "other"]`. In the rare that user dataframe has columns name `df.columns.name == [0]`, this would otherwise cause error.

That said, given this issue has affected 2 different places, we may want to refactor the `df.reorder_levels` method so we can indicate whether we want to reorder the positions explicitly, similar to what @TomAugspurger mentioned here #30289.

Reply:
I don't think this will work for MultiIndex at all

Reply:
@WillAyd It works. Actually this was written with MultiIndex in mind. 
Anyway I have added tests for MultiIndex cases. Can you review?